### PR TITLE
ports/esp32/machine_sdcard.c: sdcard read/write speedup

### PR DESCRIPTION
**Pull request by ccccmagicboy**
Use the 40MHz freq for SD card clk, tested on a esp32 module, and 30% read/write speed improve.

### CODE REVIEW

**Comment by tve** [comment]
Nice! But shouldn't this be enabled via a KW arg? Also, a doc update would be in order IMHO (there's an esp32 section in machine.SDCard.rst)...

**Comment by codefreax** [comment]
Nice find!

nit: use forward slashes in the commit message

File: ports/esp32/machine_sdcard.c [hunk 1]
```diff
@@ -144,6 +144,7 @@ STATIC mp_obj_t machine_sdcard_make_new(const mp_obj_type_t *type, size_t n_args
         ARG_mosi,
         ARG_sck,
         ARG_cs,
+        ARG_bitrate,
```
Comment (tve) [inline ports/esp32/machine_sdcard.c]: Hmm, is "bitrate" an accurate term here? IIUC multiple data lines can be hooked up and thus multiple bits can be transferred per clock cycle, thus "freq" or "clockrate" would be appropriate (esp-idf uses "freq").

**Review by tve — COMMENTED** [review]
Thanks for adding a KW arg! How is a user supposed to guess that it's in kHz? (docs...)

File: ports/esp32/machine_sdcard.c [hunk 2]
```diff
@@ -193,11 +198,16 @@ STATIC mp_obj_t machine_sdcard_make_new(const mp_obj_type_t *type, size_t n_args
...
+    
+    int freq = arg_vals[ARG_bitrate].u_int;
+    
     if (is_spi) {
         sdmmc_host_t _temp_host = SDSPI_HOST_DEFAULT();
         self->host = _temp_host;
     } else {
         sdmmc_host_t _temp_host = SDMMC_HOST_DEFAULT();
+        // Use SDMMC_FREQ_HIGHSPEED = 40MHz, about 30% read/write speed improve over SDMMC_FREQ_DEFAULT = 20MHz
+        _temp_host.max_freq_khz = freq;
```
Comment (codefreax) [inline ports/esp32/machine_sdcard.c]: Now the comment is not right here anymore. Maybe just drop it and keep the info in the commit message only?

**Review by tve — COMMENTED** [review]
:+1:

File: docs/library/machine.SDCard.rst [hunk 3]
```diff
@@ -50,6 +50,8 @@ vary from platform to platform.
      - *mosi* can be used to specify an SPI mosi pin.
 
      - *cs* can be used to specify an SPI chip select pin.
+     
+     - *freq_khz* selects the SD/MMC interface frequency in KHz unit.
```
Comment (tve) [inline docs/library/machine.SDCard.rst]: Shouldn't you indicate that this arg is esp32-only?

Comment (ccccmagicboy) [reply docs/library/machine.SDCard.rst]: Is there a separtate doc section about esp32 SDcard class?

Comment (tve) [reply docs/library/machine.SDCard.rst]: I don't think so. I would just add a "Note: only supported on the ESP32" or something like that...

Comment (ccccmagicboy) [reply docs/library/machine.SDCard.rst]: OK

**Comment by dpgeorge** [comment]
Thanks for the PR.

There are a lot of places in the existing machine API that use `freq` and it's measured in Hz; eg `machine.Timer`, `machine.I2C`, and `machine.freq()`.  So for consistency I really think it's a good idea to have this new arg as `freq` in Hz as well.

And I'm not sure the new constants for set frequencies are needed: it's possible to just pass in an integer with the desired value.  This is again for consistency with other places that have a `freq` kwarg.

File: ports/esp32/machine_sdcard.c [hunk 4]
```diff
@@ -175,6 +178,8 @@ STATIC mp_obj_t machine_sdcard_make_new(const mp_obj_type_t *type, size_t n_args
         arg_vals[ARG_miso].u_obj, arg_vals[ARG_mosi].u_obj,
         arg_vals[ARG_sck].u_obj, arg_vals[ARG_cs].u_obj);
 
+    DEBUG_printf("  freq_khz=%d", arg_vals[ARG_freq_khz].u_int);
```
Comment (dpgeorge) [inline ports/esp32/machine_sdcard.c]: This debug line isn't needed.

File: ports/esp32/machine_sdcard.c [hunk 5]
```diff
@@ -193,11 +198,15 @@ STATIC mp_obj_t machine_sdcard_make_new(const mp_obj_type_t *type, size_t n_args
     self->flags = 0;
     // Note that these defaults are macros that expand to structure
     // constants so we can't directly assign them to fields.
+
+    int freq_khz = arg_vals[ARG_freq_khz].u_int;
```
Comment (dpgeorge) [inline ports/esp32/machine_sdcard.c]: For coding style, this can go down inside the "else" block below where it's used.

File: ports/esp32/machine_sdcard.c [hunk 6]
```diff
@@ -244,7 +253,8 @@ STATIC mp_obj_t machine_sdcard_make_new(const mp_obj_type_t *type, size_t n_args
 
         // Stronger external pull-ups are still needed but apparently
         // it is a good idea to set the internal pull-ups anyway.
-        // slot_config.flags = SDMMC_SLOT_FLAG_INTERNAL_PULLUP;
+        // test for 4line card
+        slot_config.flags = SDMMC_SLOT_FLAG_INTERNAL_PULLUP;
```
Comment (dpgeorge) [inline ports/esp32/machine_sdcard.c]: This looks like an independent change not related to the freq addition.  Is that right?

Comment (ccccmagicboy) [reply ports/esp32/machine_sdcard.c]: Restore it.

**Comment by ccccmagicboy** [comment]
Here some random test result on a custom build and custom esp32-wrover board below for reference: 
```python
import machine
sd = machine.SDCard(slot=1, width=4, freq=40000000)
# write: 0.088MByte/s, read: 0.925MByte/s
sd = machine.SDCard(slot=1, width=1, freq=20000000)
# write: 0.085MByte/s, read: 0.687MByte/s
```

File: ports/esp32/Makefile [hunk 7]
```diff
@@ -53,8 +53,8 @@ SDKCONFIG_H = $(BUILD)/sdkconfig.h
 
 # The git hash of the currently supported ESP IDF version.
 # These correspond to v3.3.1 and v4.0.
-ESPIDF_SUPHASH_V3 := 143d26aa49df524e10fb8e41a71d12e731b9b71d
-ESPIDF_SUPHASH_V4 := 463a9d8b7f9af8205222b80707f9bdbba7c530e1
+ESPIDF_SUPHASH_V3 := 5d34675de13aaa8a2873de730fd2b6ac71f462b1
+ESPIDF_SUPHASH_V4 := 2ce5478da5dd9e54e74acf4606a2e5b32a066b16
```
Comment (tve) [inline ports/esp32/Makefile]: looks like you need to rebase on master? or remove this change...

File: ports/esp32/machine_sdcard.c [hunk 8]
```diff
@@ -278,6 +290,15 @@ STATIC mp_obj_t sd_deinit(mp_obj_t self_in) {
...
 STATIC MP_DEFINE_CONST_FUN_OBJ_1(sd_deinit_obj, sd_deinit);
 
+STATIC mp_obj_t sd_card_print_info(mp_obj_t self_in) {
+    sdcard_card_obj_t *self = self_in;
+
+    sdmmc_card_print_info(stdout, &(self->card));
+
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(sd_card_print_info_obj, sd_card_print_info);
```
Comment (dpgeorge) [inline ports/esp32/machine_sdcard.c:300]: This is an unrelated addition and shouldn't be in this PR (it requires a separate discussion).

File: ports/esp32/machine_sdcard.c [hunk 9]
```diff
@@ -57,6 +57,10 @@
 // Forward declaration
 const mp_obj_type_t machine_sdcard_type;
 
+//fix the sdspi highspeed select error
+esp_err_t sdmmc_init_card_hs_mode_fix(sdmmc_card_t* card);
+esp_err_t sdmmc_card_init_fix(const sdmmc_host_t* config, sdmmc_card_t* card);
```
Comment (dpgeorge) [inline ports/esp32/machine_sdcard.c:62]: What are these for?

File: ports/esp32/machine_sdcard.c (hunk 8 above)
Comment (ccccmagicboy) [reply ports/esp32/machine_sdcard.c:278]: OK!

File: ports/esp32/machine_sdcard.c (hunk 9 above)
Comment (ccccmagicboy) [reply ports/esp32/machine_sdcard.c]: I think i want to try using more higher freq for the sdmmc module at that time which idf lib limited, but failed. Already remove that code.

**Comment by ccccmagicboy** [comment]
Sorry for the late reply, commits is rebased and squashed, irrelevant information has been removed, can this PR merge?@dpgeorge

**Comment by dpgeorge** [comment]
Thanks for updating, now merged in 1662a0b06f2ad58aa4a632a88d0b3836c08bded2
